### ISSUE TITLE
Docker volumes are created in /var/lib/docker/volumes even with devicemapper

### ISSUE THREAD

**Issue by rlpowell** [docker/docker#27215]
**Description**

I'm using devicemapper, and have allocated disk space to that purpose (see info output below).

Part of the reason I did this is that I don't have enough space in /var/lib/docker to store my docker volumes, but this command:

      sudo docker create -v /home/rlpowell/.stack -v /home/rlpowell/.local --name stackstore rlpowell/shell /bin/true

creates the volume in /var/lib/docker/volumes:

```
rlpowell@vrici> sudo docker volume inspect b8d5d1ec03ac6b10fe6fa90c3022613290ae726bd483383342a1037665e5866e
[
    {
        "Name": "b8d5d1ec03ac6b10fe6fa90c3022613290ae726bd483383342a1037665e5866e",
        "Driver": "local",
        "Mountpoint": "/var/lib/docker/volumes/b8d5d1ec03ac6b10fe6fa90c3022613290ae726bd483383342a1037665e5866e/_data",
        "Labels": null,
        "Scope": "local"
    }
]
```

**Describe the results you received:**

My / filling due to docker volume operations.

**Describe the results you expected:**

I expected everything docker to go into the devicemapper LV

**Output of `docker version`:**

```
Client:
 Version:         1.12.1
 API version:     1.24
 Package version: docker-common-1.12.1-28.git9a3752d.fc26.x86_64
 Go version:      go1.7.1
 Git commit:      9a3752d/1.12.1
 Built:
 OS/Arch:         linux/amd64

Server:
 Version:         1.12.1
 API version:     1.24
 Package version: docker-common-1.12.1-28.git9a3752d.fc26.x86_64
 Go version:      go1.7.1
 Git commit:      9a3752d/1.12.1
 Built:
 OS/Arch:         linux/amd64
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 77
Server Version: 1.12.1
Storage Driver: devicemapper
 Pool Name: docker-docker--pool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file:
 Metadata file:
 Data Space Used: 4.269 GB
 Data Space Total: 48.32 GB
 Data Space Available: 44.05 GB
 Metadata Space Used: 1.835 MB
 Metadata Space Total: 54.53 MB
 Metadata Space Available: 52.69 MB
 Thin Pool Minimum Free Space: 4.832 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135 (2016-09-26)
Logging Driver: journald
Cgroup Driver: systemd
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: inactive
Runtimes: oci runc
Default Runtime: oci
Security Options: seccomp selinux
Kernel Version: 4.6.0-0.rc6.git0.1.fc25.x86_64
Operating System: Fedora 26 (Server Edition)
OSType: linux
Architecture: x86_64
Number of Docker Hooks: 2
CPUs: 4
Total Memory: 2.801 GiB
Name: vrici.digitalkingdom.org
ID: JIIS:TCH7:ZYXV:M2KK:EXQH:GZPY:OAPY:2DJF:SE7A:UZBO:A3PX:NUWF
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://registry.access.redhat.com/v1/
Insecure Registries:
 127.0.0.0/8
Registries: registry.access.redhat.com (secure), docker.io (secure)
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

This is on a KVM/Qemu virtual host on a physical box in my basement.

**Comment by tonistiigi** [comment]
Storage driver is only used for storing image layers. Not volumes, configuration files etc. If you want to switch all storage to a different path/drive use the `--graph` option.

**Comment by rlpowell** [comment]
Thanks for the workaround.  The current situation seems a bit anti-POLS to me, though; I told docker to use devicemapper, I expect it to do so.

**Comment by rlpowell** [comment]
Wait, hold on.  If I'm reading correctly --graph is an option I give to the daemon that points it at some other place in the file tree, instead of /var/lib/docker ; that doesn't help me at all.  The whole issue is that I allocated a bunch of space to the unformatted docker LV for docker to use; there is no place in my filesystem tree to point to.

I guess I'll just format the LV with a regular file system and go back to using overlay.

**Comment by ssuvasanth** [comment]
Storage driver is not used for managing the data volumes. As mentioned by @tonistiigi, they are used only to manage the image layers. 
Data volumes are managed by volume drivers. By default, the local volume driver is used which keeps the volumes under `/var/lib/docker/volumes`.  If you intend to use an unused LV for volumes, mount it at `/var/lib/docker/volumes`.

**Comment by shishir-a412ed** [comment]
@rlpowell I agree this is a valid usecase. You don't want to exhaust your disk space under the root tree `/`, by creating a lot of volumes which are stored under `/var/lib/docker/volumes`.

One solution to your problem is `docker volume plugins`. I don't know if you have used that before.
I wrote a `docker-lvm-plugin` which seems to be more optimal for your usecase. You can check it out at the below link:

https://github.com/projectatomic/docker-lvm-plugin

With this plugin you can have a separate volume group (VG) dedicated for your volumes.  You can carve out volumes and they will come out of this volume group and not stored under `/var/lib/docker/volumes`. You can also create a thin pool and provision thin volumes, take snapshots of thin and regular volumes. 

PS: thin volumes are more space friendly since they are COW and don't take up space unless you write on the volume.

Shishir

**Comment by thaJeztah** [comment]
@rlpowell one reason to use _volumes_ is to prevent the overhead that layered file systems give; in your case (I see you're using an anonymous volume), is there a reason to use a volume at all? Omitting the volume will just use the containers filesystem (which is on devicemapper).

Other than that, I think the question here is answered; volumes are separate from the graphdriver / containers filesystem (and for a reason); if you need a specific type of volume (for example, a volume that uses devicemapper), a volume driver plugin is the right approach.

**Comment by docwhat** [comment]
I'm finding this a bit confusing... I think my problem falls under this issue:

I have a huge docker image (22G) built with:

```
FROM scratch
COPY data/ /data/
VOLUME /data/
```

I cannot `docker create --name=mydata --read-only myimage:latest nop` because it copies everything out to `/var/lib/docker/volumes`?

I expected that it would leave everything in the devicemapper lvm-direct storage space and then I could use it via `docker run --name=mytests --volumes-from=mydata ...` without ending up with duplicate copy of my 22G on the host's root filesystem.

The `mydata` is test data (relatively unchanging) and I want to access it from `mytests` so I can run tests using the corpus of test data.

What's the right approach to do this then?  Does something like the `docker-lvm-plugin` mentioned above even come into play?

**Comment by cpuguy83** [comment]
I do not believe we populate volumes via `volumes-from`.
It's also not copying your entire image into the volume, it's copying the data from the image at requested volume path into the volume.

**Comment by docwhat** [comment]
> I do not believe we populate volumes via volumes-from.

In my testing, I haven't even gotten to the `--volumes-from=` part... I'm stuck on `docker create...`.  

It runs out of space as it copies everything (all the data is in the `VOLUME` directory) into `/var/lib/docker/volumes/2fb0353bcc6aa90db1638e85baa0ab42eb988db2ec94fa53fb18e8e0f0003125`.  I did an `ls` in there and it looks like the `/data/` directory from my image.

So this isn't expected behavior then?

**Comment by cpuguy83** [comment]
@docwhat If you have data in `/data` and have a `VOLUME /data` then yes it is expected to copy everything from `/data`. If you do not want this behavior, then you can either use the `nocopy` flag with a named volume (e.g. `docker run -v mydata:/data:nocopy`) or remove all data in `/data` in the image.

**Comment by docwhat** [comment]
`:nocopy`? I haven't seen that before.

**Comment by docwhat** [comment]
...And that'd be too late, since the `docker create` for the `mydata` container has already copied everything to `/var/lib/docker/volumes/xxxx`.

**Comment by docwhat** [comment]
Just to be clear... given this `run.sh` script...

```
#!/bin/bash

set -euxo pipefail

cat <<EOF > Dockerfile
FROM scratch
COPY ./ /data/
VOLUME /data/
CMD ["nop"]
EOF

date > file1
date --utc > file2

docker build -t data .
docker create --name=mydata data nop
docker inspect mydata | jq '.[0].Mounts[0]'
dir=$(docker inspect mydata | jq '.[0].Mounts[0].Source' -r)

ls -al "$dir"
```

I get this output:

```
+ cat
+ date
+ date --utc
+ docker build -t data .
Sending build context to Docker daemon  5.12 kB
Step 1 : FROM scratch
 --->
Step 2 : COPY ./ /data/
 ---> af348fc85577
Removing intermediate container ad95fe7f353e
Step 3 : VOLUME /data/
 ---> Running in e0a8d052fcb7
 ---> 35f83caae681
Removing intermediate container e0a8d052fcb7
Step 4 : CMD nop
 ---> Running in 635a40e3b984
 ---> 5583ecb7835c
Removing intermediate container 635a40e3b984
Successfully built 5583ecb7835c
+ docker create --name=mydata data nop
daa3b7f88bbde3139d8a598bb98980400e2abaa4bdd4851a9fd38399cb19693a
+ docker inspect mydata
+ jq '.[0].Mounts[0]'
{
  "Name": "5f0e158aa1acf0621c4d05af44fd804be7c0aa7d6c0983430b2ada81e81e5b75",
  "Source": "/var/lib/docker/volumes/5f0e158aa1acf0621c4d05af44fd804be7c0aa7d6c0983430b2ada81e81e5b75/_data",
  "Destination": "/data",
  "Driver": "local",
  "Mode": "",
  "RW": true,
  "Propagation": ""
}
++ docker inspect mydata
++ jq '.[0].Mounts[0].Source' -r
+ dir=/var/lib/docker/volumes/5f0e158aa1acf0621c4d05af44fd804be7c0aa7d6c0983430b2ada81e81e5b75/_data
+ ls -al /var/lib/docker/volumes/5f0e158aa1acf0621c4d05af44fd804be7c0aa7d6c0983430b2ada81e81e5b75/_data
total 16
drwxr-xr-x 2 root root  60 Nov 14 15:55 .
drwxr-xr-x 3 root root  18 Nov 14 15:55 ..
-rw-r--r-- 1 root root  54 Nov 14 15:55 Dockerfile
-rw-r--r-- 1 root root  29 Nov 14 15:55 file1
-rw-r--r-- 1 root root  29 Nov 14 15:55 file2
-rwxr-xr-x 1 root root 328 Nov 14 15:55 run.sh
```

1. This is expected?
2. And there is no way to prevent the `/var/lib/docker/volumes/5f0e158aa1acf0621c4d05af44fd804be7c0aa7d6c0983430b2ada81e81e5b75` directory from being created and just letting it export directly from the LVM device/snapshot?

I'm expecting the answers to be "yes" and "no, that's how it works", but I want to be certain.

**Comment by cpuguy83** [comment]
Yes that's correct.
It sounds like you don't want to do `VOLUME /data` in the dockerfile.

**Comment by docwhat** [comment]
Okay, thanks for clearing that confusion up! I didn't expect to suddenly see `/var/lib/docker` filling up because I didn't understand what was really going on under the covers.

How would you suggest dealing with huge amounts of (versioned, though infrequently changing) data like this?

**Comment by cpuguy83** [comment]
@docwhat Generally I wouldn't put data into an image.

**Comment by jantman** [comment]
I know this has been closed. But there's a LOT of information out there about docker LVM setup... and it seems that a LOT of people have made the same assumption that I have... that if I set up Docker for LVM thinpooldev, it will store its "stuff" there. We'd been running that way for almost two months before discovering that our data volumes (which we use to allow other containers to access data from a given container, and for backups) are ending up on the rootfs.

Would it be possible to get a clarification of this in the docs, somewhere obvious?
